external/opensocdebug/hardware/modules/stm/modtestlib.py

```diff
@@ -0,0 +1,469 @@
+        yield receive_packet(dut, response)
+        
+        # if the module returned a success, everything is fine
+        if ((response[2] & 0x3C00) >> 10) == CONST_PACKET_SUBTYPE_RESP_READ_REG_SUCCESS_64:
+            print("Read process was successful!")
+        # if it returned an error, output that error
+        elif ((response[2] & 0x3C00) >> 10) == CONST_PACKET_SUBTYPE_RESP_READ_REG_ERROR:
+            print("An error occurred during the read process!")
+        # otherwise something else went wrong
+        else:
+            print("An unidentified error occurred!")
+            
+    elif reg_width == 128:
+        # transmit the read request
+        yield send_packet(dut, destination, source, CONST_PACKET_TYPE_REG, CONST_PACKET_SUBTYPE_REQ_READ_REG_128,  address)
+        
+        # read in the response
+        yield receive_packet(dut, response)
+        
+        # if the module returned a success, everything is fine
+        if ((response[2] & 0x3C00) >> 10) == CONST_PACKET_SUBTYPE_RESP_READ_REG_SUCCESS_128:
+            print("Read process was successful!")
+        # if it returned an error, output that error
+        elif ((response[2] & 0x3C00) >> 10) == CONST_PACKET_SUBTYPE_RESP_READ_REG_ERROR:
+            print("An error occurred during the read process!")
+        # otherwise something else went wrong
+        else:
+            print("An unidentified error occurred!")
+            
+    else:
+        # display an error message if an unsupported register width was chosen
+        raise TestFailure("Invalid register width parameter")
+    
+    for _ in range(3):
+        yield RisingEdge(dut.clk)
+
+    
+@cocotb.coroutine        
+def ModuleSubTest_STM(dut, test_id):
+    """function which takes a second parameter aside from the device under test to specify which functionality of the debug module should be tested and which tells the user if the correct values are stored inside the base registers of a debug module
+        
+        Args:
+        dut:            device under test.
+        test_id:        ID indicating which functionality should be tested.
+        
+        Returns:
+        -
+        
+    """
+    
+    # other tests could potentially be added
+    if test_id == CONST_BASE_REG_TEST:
+            
+        response = []
+        
+        #============================================================================================================Test MOD_ID
+        
+        # specify how the result should look like (has to be changed to defining a single value instead of a list)
+        expected_payload = []
+        expected_payload.append(1)
+        
+        #             dut, destination, source, word width, reg address, list for return value
+        yield RegRead(dut, 1,           0,      16,         0,           response)
+    
+        #                     dut, response, destination(dut), source(simulated), word width, expected payload, success variable
```

Comment:
Comment looks a bit odd. It is a great idea to see what the parameters mean, but probably its better to use the parameters as keyword parameters then:

    yield RegRead(dut, destination=1, ...)

For Python 3 I would suggest to use keyword only functions:

    def RegRead(*,destination,...)

This avoids errors in calling the functions.